File: week4/lee/체육복.js
function solution(n, lost, reserve) {
    var answer = 0;
    lost.sort()
    reserve.sort()

    // 잃어버린 사람과 여분 있는 사람이 같은 경우 필터
    let l = lost.filter((v)=>{
        return !reserve.includes(v)
    })
    let r = reserve.filter((v)=>{
       return !lost.includes(v);
    });


    const a = l.filter((v)=>{
        let extra = r.find((rr)=>{//빌려줄 수 있는 경우
            return Math.abs(v-rr) <=1;
        });
        if(!extra) return true; // 빌려줄수 없는 경우 a에 추가
        r = r.filter((q)=>q!==extra) // r배열 업데이트
    })
    
    answer = n-a.length //전체 - 못빌린 사람의 수 
    return answer;
}

File: week4/eunbeen/체육복.java
import java.util.*;
class Solution {
    public int solution(int n, int[] lost, int[] reserve) {
        int answer = 0;
        Queue<Integer> queue1 = new LinkedList();
        Queue<Integer> queue2 = new LinkedList();
        
        Arrays.sort(lost);
        Arrays.sort(reserve);
        
        for(int i=0; i<lost.length; i++){
            for(int j=0; j<reserve.length; j++){
                if(lost[i] == reserve[j]){
                    lost[i] = -1;
                    reserve[j] = -1;
                }
            }
        }
        
        for(int i : lost){
            if(i == -1) continue;
            queue1.add(i);
        }
        
        for(int i : reserve){
            if(i == -1) continue;
            queue2.add(i);
        }
        
        if(queue2.size() == 0){
            return n-queue1.size();
        }
        
        System.out.println(queue1 + " " + queue2);
        while(queue2.size() > 0){
            int num = queue2.remove();
            
            while(queue1.size() > 0 && queue1.peek() < (num+2)){
                int num2 = Math.abs(queue1.remove() - num);
                if(num2 != 1){
                    n--;
                }
                else {
                    break;
                }
            }
            
        }
         if(queue1.size() > 0){
            return n-queue1.size();
        }
        System.out.println(queue1 + " " + queue2);
        
        return n;
    }
}